{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6044de55-9594-4d95-bccd-866ba6cb9d0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import library\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "import random\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93e292dd-2394-4692-a219-1f407a4bb920",
   "metadata": {},
   "outputs": [],
   "source": [
    "# setting\n",
    "plt.rcParams['font.family']= 'sans-serif'\n",
    "plt.rcParams['font.sans-serif'] = ['Arial']\n",
    "plt.rcParams['font.size'] = 18\n",
    "plt.rcParams[\"figure.dpi\"] = 200"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4411cb38-ca82-4817-88f9-a67389ca7e28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of diel oscillating genes (DOGs)\n",
    "DOGs_list = []\n",
    "hm_list = [\"H3K4me2\",\"H3K4me3\",\"H2A.Z\",\"H2Aub\"]\n",
    "for hm in hm_list:\n",
    "    file_path = f\"../data/list_of_DOGs/{hm}_DOGs.bed\"\n",
    "    DOGs = pd.read_csv(file_path,sep=\"\\t\",header=None).iloc[:,3].values\n",
    "    DOGs_list.append(DOGs)\n",
    "\n",
    "file_path = \"../data/list_of_DOGs/mRNA_DOGs.txt\"\n",
    "DOGs = pd.read_csv(file_path).columns.values\n",
    "DOGs_list.append(DOGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbb2e6b4-e411-4b6e-85d8-5509dc4afeb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fig. 5c\n",
    "hm_list = [\"H3K4me2\",\"H3K4me3\",\"H2A.Z\",\"H2Aub\",\"mRNA\"]\n",
    "hm_list_with_n = [\"H3K4me2 DOGs\\n(n=781)\",\"H3K4me3 DOGs\\n(n=1,180)\",\"H2A.Z DOGs\\n(n=345)\",\"H2Aub DOGs\\n(n=29)\",\"mRNA DOGs\\n(n=3,297)\"]\n",
    "N = 27443 # Number of protein-coding genes\n",
    "kinds_of_DOGs = 5\n",
    "\n",
    "M = np.zeros([kinds_of_DOGs,kinds_of_DOGs]) # Actual Number of overlapping genes\n",
    "E = np.zeros([kinds_of_DOGs,kinds_of_DOGs]) # Expected number of overlapping genes\n",
    "F = np.zeros([kinds_of_DOGs,kinds_of_DOGs]) # Fisher's exact test\n",
    "J = np.zeros([kinds_of_DOGs,kinds_of_DOGs]) # Jaccard index\n",
    "\n",
    "for i,X1 in enumerate(DOGs_list):\n",
    "    for j,X2 in enumerate(DOGs_list):\n",
    "        n1 = len(X1)\n",
    "        n2 = len(X2)\n",
    "        m = len(set(X1) & set(X2))\n",
    "        e = n1*n2/N\n",
    "        r = m/e\n",
    "        M[i][j] = m\n",
    "        E[i][j] = e\n",
    "        # print(f\"{hm_list[i]}-{hm_list[j]} : {n1},{n2},{m},{e},{r}\")\n",
    "        \n",
    "        # Fisher's exact test\n",
    "        f = scipy.stats.fisher_exact(np.array([[m, n1-m],[n2-m,N-(n1+n2-m)]]),alternative='greater')\n",
    "        F[i][j] = f.pvalue * 10\n",
    "        \n",
    "        # Jaccard index\n",
    "        J[i][j] = m/(n1+n2-m)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3.5,3.5))\n",
    "for i in range(kinds_of_DOGs):\n",
    "    for j in range(0,i):\n",
    "        if i==j:\n",
    "            continue\n",
    "        # plot\n",
    "        ax.scatter(j,kinds_of_DOGs-i-1,s=J[i][j]*2000,c=\"grey\")\n",
    "        # show results of Fisher's exact test\n",
    "        text = None\n",
    "        if F[i][j] < 0.05:\n",
    "            text = f\"*{M[i][j]:.0f}\\n ({E[i][j]:.0f})\"\n",
    "        else:\n",
    "            text = f\"{M[i][j]:.0f}\\n ({E[i][j]:.0f})\"\n",
    "        ax.text(j,kinds_of_DOGs-i-1,text,size=11,fontname=\"arial\",fontweight=\"bold\")\n",
    "\n",
    "\n",
    "        \n",
    "# format figure\n",
    "ax.set_xlim(-0.5,kinds_of_DOGs-1.5)\n",
    "ax.set_ylim(-0.5,kinds_of_DOGs-1.5)\n",
    "ax.set_xticks(np.arange(kinds_of_DOGs-1))\n",
    "ax.set_yticks(np.arange(kinds_of_DOGs-1))\n",
    "ax.set_xticklabels(hm_list_with_n[:4],rotation=90,fontsize=10,fontname=\"arial\",fontweight=\"bold\")\n",
    "ax.set_yticklabels(hm_list_with_n[::-1][:4],fontsize=10,fontname=\"arial\",fontweight=\"bold\")\n",
    "plt.gca().spines['right'].set_visible(False)\n",
    "plt.gca().spines['top'].set_visible(False)\n",
    "plt.gca().spines['left'].set_visible(False)\n",
    "plt.gca().spines['bottom'].set_visible(False)\n",
    "plt.grid()\n",
    "ax.set_axisbelow(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "666cf8b2-fce2-48f6-ba5d-68c0e6170d2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# H3K4me2 DOGs with different peak time\n",
    "H3K4me2_DOGs_peak_time_list = []\n",
    "for i in range(4):\n",
    "    file = f\"../data/list_of_DOGs/H3K4me2_DOGs_ZT{6*i}-peak.bed\"\n",
    "    genes = pd.read_csv(file,sep=\"\\t\",header=None).iloc[:,3].values\n",
    "    H3K4me2_DOGs_peak_time_list.append(genes)\n",
    "    print(f\"ZT{6*i:<2} : {len(genes):>3} genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c018cf3-54df-4af0-b239-8ff1b46accd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fig. 5d\n",
    "# Time-course ChIP-seq data\n",
    "file_path = \"../data/ChIP-seq/Time-course-ChIP_WT_rep1.rpkm.tsv\"\n",
    "df_chip = pd.read_csv(file_path, sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2ffa8bf-6afe-4fb9-9149-818b7b9e813a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Time-course RNA-seq data\n",
    "file_path = \"../data/RNA-seq/Time-course-RNA_WT_rep1.rpkm.tsv\"\n",
    "df_rna = pd.read_csv(file_path, sep=\"\\t\", index_col=0)\n",
    "\n",
    "# mean of three replicates\n",
    "df_rna_mean = pd.DataFrame({f\"WT_mRNA_ZT{6*i}\":df_rna.iloc[:,3*i:3*(i+1)].mean(axis=1) for i in range(4)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "969e3df9-e488-4b91-893e-7f02ad2b242a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_chip.merge(df_rna_mean,left_index=True,right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1715f398-0c8b-43bd-837c-87ee8222b95e",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_list = [\"H3\",\"H3K4me2\",\"H3K4me3\",\"H2A.Z\",\"H2Aub\",\"mRNA\"]\n",
    "color_list = [\"tab:grey\",\"tab:orange\",\"tab:cyan\",\"tab:red\",\"tab:green\",\"tab:purple\"]\n",
    "\n",
    "hm_n = 6\n",
    "zt_n = 4\n",
    "\n",
    "fig, ax = plt.subplots(hm_n,zt_n,figsize=(18,18),sharex=\"col\",sharey=\"row\",dpi=300)\n",
    "\n",
    "for i in range(hm_n):\n",
    "    hm = hm_list[i]\n",
    "    c = color_list[i]\n",
    "    for j in range(zt_n):\n",
    "        mask = df_all.index.isin(H3K4me2_DOGs_peak_time_list[j])\n",
    "        X = df_all.iloc[:,4*i:4*(i+1)].loc[mask]\n",
    "        \n",
    "        # add ZT0 as ZT24\n",
    "        X[f\"WT_{hm}_ZT24\"] = X.iloc[:,0]\n",
    "        X.columns = np.arange(zt_n+1)\n",
    "        \n",
    "        # standardization\n",
    "        std_X = X.apply(lambda x: (x - x[:4].mean())/x[:4].std(ddof=1),axis=1)\n",
    "\n",
    "        # plot\n",
    "        ax[i][j].plot(std_X.T,c=c,alpha=0.1,linewidth=1)\n",
    "        ax[i][j].plot(std_X.mean(),c=\"k\",alpha=1,linewidth=2,linestyle=\"--\",\n",
    "                      marker=\"o\")\n",
    "        \n",
    "for j in range(4):\n",
    "    ax[0][j].set_xticks(np.arange(zt_n+1))\n",
    "    ax[0][j].set_xticklabels([\"ZT0\",\"ZT6\",\"ZT12\",\"ZT18\",\"(ZT0)\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f24f40d-bb9b-41ef-bfdd-02efea58f73c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fig. 5e\n",
    "N = 6\n",
    "\n",
    "r_list = []\n",
    "hm_list = [\"H3\",\"H3K4me2\",\"H3K4me3\",\"H2A.Z\",\"H2Aub\",\"mRNA\"]\n",
    "hm_n = 6\n",
    "\n",
    "mask = df_all.index.isin(DOGs_list[0]) # H3K4me2 DOGs (n=781)\n",
    "\n",
    "for i in range(hm_n):\n",
    "    if i == 1:\n",
    "        continue\n",
    "        \n",
    "    hm = hm_list[i]\n",
    "    \n",
    "    df_tmp1 = df_all.iloc[:,4:8][mask]\n",
    "    df_tmp2 = df_all.iloc[:,4*i:4*(i+1)][mask]\n",
    "  \n",
    "    # Spearman's R \n",
    "    R = []\n",
    "    for j in range(len(df_tmp1)):\n",
    "        x = df_tmp1.iloc[j,:]\n",
    "        y = df_tmp2.iloc[j,:]\n",
    "        R.append(pearsonr(x,y)[0])\n",
    "    r_list.append(R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48491356-1ea6-4229-a410-204ceac800c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot\n",
    "color_list = [\"tab:grey\",\"tab:cyan\",\"tab:red\",\"tab:green\",\"tab:purple\"]\n",
    "fig, ax = plt.subplots(1,5,figsize=(22,4))\n",
    "\n",
    "for i in range(5):\n",
    "    ax[i].hist(r_list[i],bins=20,range=(-1,1),color=color_list[i],ec='black')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b46f9e0-a90e-4428-a100-ddae0e186fa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fig. 5f\n",
    "# Time-course ChIP-seq data in WT and atxr3\n",
    "file_path = \"../data/ChIP-seq/Time-course-ChIP_WT_atxr3_rep1.rpkm.tsv\"\n",
    "df_chip = pd.read_csv(file_path, sep=\"\\t\", index_col=0)\n",
    "\n",
    "# Time-course mRNA-seq data in WT and atxr3\n",
    "file_path = \"../data/RNA-seq/Time-course-RNA_WT_atxr3.rpkm.tsv\"\n",
    "df_rna = pd.read_csv(file_path, sep=\"\\t\", index_col=0)\n",
    "# mean of three replicates\n",
    "gt_list = [\"WT\",\"atxr3\"]\n",
    "df_rna_mean = np.sqrt(pd.DataFrame({f\"{gt_list[i]}_mRNA_ZT{6*j}\":df_rna.iloc[:,12*i+3*j:12*i+3*(j+1)].mean(axis=1) for i in range(2) for j in range(4)}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdc53d90-7796-446b-adda-c9ca69c6ca5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_chip.iloc[:,:16].merge(df_rna_mean.iloc[:,:4],left_index=True,right_index=True).merge(\n",
    "    df_chip.iloc[:,16:],left_index=True,right_index=True).merge(\n",
    "    df_rna_mean.iloc[:,4:],left_index=True,right_index=True)\n",
    "df_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3624ca6-6d04-476d-833c-58afe57e16f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_list = [\"H3\",\"H3K4me2\",\"H3K4me3\",\"H2A.Z\",\"mRNA\"]\n",
    "c_list = [\"k\",\"tab:orange\",\"tab:cyan\",\"tab:red\",\"tab:purple\"]\n",
    "\n",
    "max_list = [6,10,20,5,5]\n",
    "min_list = [-6,-30,-40,-10,-5]\n",
    "\n",
    "mask = df_chip.index.isin(DOGs_list[0]) # H3K4me2 DOGs\n",
    "N = mask.sum()\n",
    "\n",
    "# H3K4me2 non-DOGs\n",
    "random.seed(1)\n",
    "mask_nonCOGs = df_chip.index.isin(random.sample(list(df_chip.index[~mask]),N))\n",
    "#print(N,df_chip[mask_nonCOGs].index[:3])\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(1,5,figsize=(25,5))\n",
    "for i in range(5):\n",
    "    hm = hm_list[i]\n",
    "    ser_wt = df_all.iloc[:,4*i:4*(i+1)].max(axis=1) - df_all.iloc[:,4*i:4*(i+1)].min(axis=1)\n",
    "    ser_mt = df_all.iloc[:,4*(i+5):4*(i+6)].max(axis=1) - df_all.iloc[:,4*(i+5):4*(i+6)].min(axis=1)\n",
    "\n",
    "    # H3K4me2 DOGs\n",
    "    a = ser_wt[mask]\n",
    "    b = ser_mt[mask]\n",
    "    \n",
    "    # H3K4me2 non-DOGs\n",
    "    c = ser_wt[mask_nonCOGs]\n",
    "    d = ser_mt[mask_nonCOGs]\n",
    "\n",
    "    # Amplitude change (atxr3 - WT)\n",
    "    e = b-a\n",
    "    f = d-c\n",
    "\n",
    "    # plot\n",
    "    sns.violinplot({\"1\":f,\n",
    "                    \"2\":e},\n",
    "                   palette = [\"tab:grey\",c_list[i]],\n",
    "                   fill=False,\n",
    "                   cut=0,\n",
    "                   ax=ax[i],\n",
    "                   inner_kws=dict(box_width=10,whis_width=2,),\n",
    "               )\n",
    "\n",
    "    # format figure\n",
    "    ax[i].set_ylim(min_list[i],max_list[i])\n",
    "    ax[i].spines['right'].set_visible(False)\n",
    "    ax[i].spines['top'].set_visible(False)\n",
    "    ax[i].hlines(0,-0.5,1.5,color=\"black\",linestyle=\"--\",zorder=0)\n",
    "    ax[i].set_xlim(-0.5,1.5)\n",
    "    ax[i].set_xticks([0,1])\n",
    "    ax[i].set_xticklabels([\"H3K4me2\\nnon-DOGs\",\"H3K4me2\\nDOGs\"])\n",
    "    ax[i].set_title(hm)\n",
    "    \n",
    "    # Mann-Whitney U test\n",
    "    p1 = stats.mannwhitneyu(f, e, alternative=\"two-sided\")[1]\n",
    "    print(f\"{hm:10s} p={p1:.2e}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
